File: Go/Day22.go
package main

import AH "./adventhelper"

import (
	"strconv"
)

func playHand(player1 []int, player2 []int) ([]int, []int) {
	if len(player1) == 0 || len(player2) == 0 {
		return player1, player2
	}

	if player1[0] > player2[0] {
		return append(player1[1:], player1[0], player2[0]), player2[1:]
	} else {
		return player1[1:], append(player2[1:], player2[0], player1[0])
	}
}

func playGame(player1 []int, player2 []int) ([]int, []int) {
	if len(player1) == 0 || len(player2) == 0 {
		return player1, player2
	}

	new1, new2 := playHand(player1, player2)
	return playGame(new1, new2)
}

func score(deck []int) (total int) {
	m := len(deck)
	total = 0
	for i, v := range deck {
		total += (m-i) * v
	}
	return
}

func playHand2(deck1 []int, deck2 []int) ([]int, []int) {
	if len(deck1) == 0 || len(deck2) == 0 {
		return deck1, deck2
	}

	c1, c2 := deck1[0], deck2[0]

	if c1 < len(deck1) && c2 < len(deck2) { // we are playing a sub game
		// there is some shared memory shenanigans that I don't fully understand
		// so we are going to make a copy of the slices to avoid the unpleasantness
		subDeck1 := make([]int, c1)
		subDeck2 := make([]int, c2)
		copy(subDeck1, deck1[1:(c1+1)])
		copy(subDeck2, deck2[1:(c2+1)])
		_, _, p1Wins := playGame2(subDeck1, subDeck2)
		if (p1Wins) {
			return append(deck1[1:], deck1[0], deck2[0]), deck2[1:]
		} else {
			return deck1[1:], append(deck2[1:], deck2[0], deck1[0])
		}
	} else {
		if c1 > c2 {
			return append(deck1[1:], deck1[0], deck2[0]), deck2[1:]
		} else {
			return deck1[1:], append(deck2[1:], deck2[0], deck1[0])
		}
	}
}

func seenBefore(cur []int, prev [][]int) bool {
	for _, arr := range prev {
		if len(cur) != len(arr) {
			continue
		}
		ok := true
		for i := 0; i < len(cur); i++ {
			if cur[i] != arr[i] {
				ok = false
				break
			}
		}
		if ok {
			return true
		}
	}
	return false
}

func playGame2(deck1 []int, deck2 []int) ([]int, []int, bool) {
	prevDeck1 := [][]int{}
	prevDeck2 := [][]int{}
	for {
		if len(deck1) == 0 || len(deck2) == 0 {
			return deck1, deck2, (len(deck2) == 0)
		}

		if seenBefore(deck1, prevDeck1) && seenBefore(deck2, prevDeck2) {
			return deck1, deck2, true
		}

		prevDeck1 = append(prevDeck1, deck1)
		prevDeck2 = append(prevDeck2, deck2)

		deck1, deck2 = playHand2(deck1, deck2)
	}
}

func main() {
	ss, _ := AH.ReadStrFile("../input/input22.txt")
	player1 := false
	deck1 := []int{}
	deck2 := []int{}
	for _, s := range ss {
		if AH.FirstRune(s) == rune('P') {
			player1 = !player1
			continue
		}

		n, _ := strconv.Atoi(s)

		if player1 {
			deck1 = append(deck1, n)
		} else {
			deck2 = append(deck2, n)
		}
	}

	newDeck1, newDeck2 := playGame(deck1, deck2)
	part1 := 0
	if (len(newDeck1) > 0) {
		part1 = score(newDeck1)
	} else {
		part1 = score(newDeck2)
	}

	newDeck1, newDeck2, _ = playGame2(deck1, deck2)
	part2 := 0
	if (len(newDeck1) > 0) {
		part2 = score(newDeck1)
	} else {
		part2 = score(newDeck2)
	}

	AH.PrintSoln(22, part1, part2)

	return
}

File: C++/Day22.cpp
#include "AH.h"

namespace Day22
{
	typedef std::deque<uint64_t> Deck;

	class Game
	{
	public:
		Game(const Deck& d1, const Deck& d2);
		uint64_t score() const;
		void playGame1();
		bool playGame2();
		
	private:
		void playHand1();
		void playHand2();
		uint64_t deckScore(const Deck& d) const;
		Deck d1;
		Deck d2;
	};

	Game::Game(const Deck& d1, const Deck& d2) : d1(d1), d2(d2) {}

	void Game::playHand1()
	{
		if (d1.size() == 0 || d2.size() == 0)
			return;

		auto f1 = d1.front();
		auto f2 = d2.front();
		if (d1[0] > d2[0])
		{
			d1.pop_front();  d1.push_back(f1);  d1.push_back(f2);
			d2.pop_front();
		}
		else
		{
			d1.pop_front();    
			d2.pop_front();  d2.push_back(f2);  d2.push_back(f1);
		}
	}

	void Game::playGame1()
	{
		while ((d1.size() > 0) && (d2.size() > 0))
			playHand1();

		return;
	}

	uint64_t Game::deckScore(const Deck& d) const
	{
		auto m = d.size();
		uint64_t total = 0;
		for (uint64_t i = 0; i < m; ++i)
			total += (m-i) * d[i];

		return total;
	}

	uint64_t Game::score() const
	{
		if (d1.size() > 0)
			return deckScore(d1);
		else
			return deckScore(d2);
	}

	void Game::playHand2()
	{
		if (d1.size() == 0 || d2.size() == 0)
			return;		

		auto f1 = d1.front();
		auto f2 = d2.front();

		if ((f1 < d1.size()) && (f2 < d2.size()))
		{
			auto d1Copy = d1;
			d1Copy.pop_front();
			d1Copy.resize(f1);
			auto d2Copy = d2;
			d2Copy.pop_front();
			d2Copy.resize(f2);

			Game subG(d1Copy, d2Copy);
			auto p1Wins = subG.playGame2();
			
			if (p1Wins)
			{
				d1.pop_front();  d1.push_back(f1);  d1.push_back(f2);
				d2.pop_front();
			}
			else
			{
				d1.pop_front();    
				d2.pop_front();  d2.push_back(f2);  d2.push_back(f1);
			}
		} else {
			if (f1 > f2)
			{
				d1.pop_front();  d1.push_back(f1);  d1.push_back(f2);
				d2.pop_front();
			}
			else
			{
				d1.pop_front();    
				d2.pop_front();  d2.push_back(f2);  d2.push_back(f1);
			}	
		}
		return;
	}

	bool seenBefore(const Deck& cur, const std::vector<Deck>& prev)
	{
		for (auto dq : prev)
		{
			if (dq.size() != cur.size())
				continue;
			bool ok = true;
			for (size_t i = 0; i < dq.size(); ++i)
			{
				if (cur[i] != dq[i])
				{
					ok = false;
					break;
				}
			}
			if (ok)
				return true;
		}
		return false;
	}

	bool Game::playGame2()
	{
		std::vector<Deck> prevD1s;
		std::vector<Deck> prevD2s;

		while (true)
		{
			if ((d1.size() == 0) || (d2.size() == 0))
				return (d2.size() == 0);

			if ( seenBefore(d1, prevD1s) || seenBefore(d2, prevD2s) )
				return true;

			prevD1s.push_back(d1);
			prevD1s.push_back(d2);

			playHand2();
		}
	}

	int Run(const std::string& filename)
	{
		auto inputLines = AH::ReadTextFile(filename);
		bool player1 = false;
		Deck deck1;
		Deck deck2;

		for (auto s: inputLines)
		{
			if (s.front() == 'P')
			{
				player1 = !player1;
				continue;
			}

			if (s.length() == 0)
				continue;

			uint64_t n = std::stoi(s);

			if (player1)
				deck1.push_back(n);
			else
				deck2.push_back(n);
		}

		Game g1(deck1, deck2);
		g1.playGame1();

		Game g2(deck1, deck2);
		g2.playGame2();

		AH::PrintSoln(22, g1.score(), g2.score());

		return 0;
	}

}
